File: pyzoo/zoo/models/recommendation/wide_and_deep.py
#
# Copyright 2018 Analytics Zoo Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import sys

from zoo.models.common import *
from zoo.models.recommendation import Recommender
from zoo.common.utils import callZooFunc
from zoo.pipeline.api.keras.layers import *

if sys.version >= '3':
    long = int
    unicode = str


class ColumnFeatureInfo(object):
    """
    The same data information shared by the WideAndDeep model and its feature generation part.

    Each instance could contain the following fields:
    wide_base_cols: Data of wide_base_cols together with wide_cross_cols will be fed
                    into the wide model. List of String. Default is an empty list.
    wide_base_dims: Dimensions of wide_base_cols. The dimensions of the data in
                    wide_base_cols should be within the range of wide_base_dims.
                    List of int. Default is an empty list.
    wide_cross_cols: Data of wide_cross_cols will be fed into the wide model.
                     List of String. Default is an empty list.
    wide_cross_dims: Dimensions of wide_cross_cols. The dimensions of the data in
                     wide_cross_cols should be within the range of wide_cross_dims.
                     List of int. Default is an empty list.
    indicator_cols: Data of indicator_cols will be fed into the deep model as multi-hot vectors.
                    List of String. Default is an empty list.
    indicator_dims: Dimensions of indicator_cols. The dimensions of the data in
                    indicator_cols should be within the range of indicator_dims.
                    List of int. Default is an empty list.
    embed_cols: Data of embed_cols will be fed into the deep model as embeddings.
                List of String. Default is an empty list.
    embed_in_dims: Input dimension of the data in embed_cols. The dimensions of the data in
                   embed_cols should be within the range of embed_in_dims.
                   List of int. Default is an empty list.
    embed_out_dims: The dimensions of embeddings. List of int. Default is an empty list.
    continuous_cols: Data of continuous_cols will be treated as continuous values for
                     the deep model. List of String. Default is an empty list.
    label: The name of the 'label' column. String. Default is 'label'.
    """

    def __init__(self, wide_base_cols=None, wide_base_dims=None, wide_cross_cols=None,
                 wide_cross_dims=None, indicator_cols=None, indicator_dims=None,
                 embed_cols=None, embed_in_dims=None, embed_out_dims=None,
                 continuous_cols=None, label="label", bigdl_type="float"):
        self.wide_base_cols = [] if not wide_base_cols else wide_base_cols
        self.wide_base_dims = [] if not wide_base_dims else [int(d) for d in wide_base_dims]
        self.wide_cross_cols = [] if not wide_cross_cols else wide_cross_cols
        self.wide_cross_dims = [] if not wide_cross_dims else [int(d) for d in wide_cross_dims]
        self.indicator_cols = [] if not indicator_cols else indicator_cols
        self.indicator_dims = [] if not indicator_dims else [int(d) for d in indicator_dims]
        self.embed_cols = [] if not embed_cols else embed_cols
        self.embed_in_dims = [] if not embed_in_dims else [int(d) for d in embed_in_dims]
        self.embed_out_dims = [] if not embed_out_dims else [int(d) for d in embed_out_dims]
        self.continuous_cols = [] if not continuous_cols else continuous_cols
        self.label = label
        self.bigdl_type = bigdl_type

    def __reduce__(self):
        return ColumnFeatureInfo, (self.wide_base_cols, self.wide_base_dims, self.wide_cross_cols,
                                   self.wide_cross_dims, self.indicator_cols, self.indicator_dims,
                                   self.embed_cols, self.embed_in_dims, self.embed_out_dims,
                                   self.continuous_cols, self.label)

    def __str__(self):
        return "ColumnFeatureInfo {wide_base_cols: %s, wide_base_dims: %s, wide_cross_cols: %s, " \
               "wide_cross_dims: %s, indicator_cols: %s, indicator_dims: %s, embed_cols: %s, " \
               "embed_cols: %s, embed_in_dims: %s, embed_out_dims: %s, continuous_cols: %s, " \
               "label: '%s'}" \
               % (self.wide_base_cols, self.wide_base_dims, self.wide_cross_cols,
                  self.wide_cross_dims, self.indicator_cols, self.indicator_dims,
                  self.embed_cols, self.embed_cols, self.embed_in_dims,
                  self.embed_out_dims, self.continuous_cols, self.label)

    @property
    def feature_cols(self):
        return self.wide_base_cols + self.wide_cross_cols +\
            self.indicator_cols + self.embed_cols + self.continuous_cols

    @property
    def label_cols(self):
        return [self.label]


class WideAndDeep(Recommender):
    """
    The Wide and Deep model used for recommendation.

    # Arguments
    class_num: The number of classes. Positive int.
    column_info: An instance of ColumnFeatureInfo.
    model_type: String. 'wide', 'deep' and 'wide_n_deep' are supported. Default is 'wide_n_deep'.
    hidden_layers: Units of hidden layers for the deep model.
                   Tuple of positive int. Default is (40, 20, 10).
    """

    def __init__(self, class_num, column_info, model_type="wide_n_deep",
                 hidden_layers=[40, 20, 10], bigdl_type="float"):
        assert len(column_info.wide_base_cols) == len(column_info.wide_base_dims), \
            "size of wide_base_columns should match"
        assert len(column_info.wide_cross_cols) == len(column_info.wide_cross_dims), \
            "size of wide_cross_columns should match"
        assert len(column_info.indicator_cols) == len(column_info.indicator_dims), \
            "size of wide_indicator_columns should match"
        assert len(column_info.embed_cols) == len(column_info.embed_in_dims) \
            == len(column_info.embed_out_dims), "size of wide_indicator_columns should match"

        self.class_num = int(class_num)
        self.wide_base_dims = column_info.wide_base_dims
        self.wide_cross_dims = column_info.wide_cross_dims
        self.indicator_dims = column_info.indicator_dims
        self.embed_in_dims = column_info.embed_in_dims
        self.embed_out_dims = column_info.embed_out_dims
        self.continuous_cols = column_info.continuous_cols
        self.model_type = model_type
        self.hidden_layers = [int(unit) for unit in hidden_layers]
        self.bigdl_type = bigdl_type
        self.model = self.build_model()
        super(WideAndDeep, self).__init__(None, self.bigdl_type,
                                          self.model_type,
                                          self.class_num,
                                          self.hidden_layers,
                                          self.wide_base_dims,
                                          self.wide_cross_dims,
                                          self.indicator_dims,
                                          self.embed_in_dims,
                                          self.embed_out_dims,
                                          self.continuous_cols,
                                          self.model)

    def build_model(self):
        wide_dims = sum(self.wide_base_dims) + sum(self.wide_cross_dims)
        input_wide = Input(shape=(wide_dims,))
        input_ind = Input(shape=(sum(self.indicator_dims),))
        input_emb = Input(shape=(len(self.embed_in_dims),))
        input_con = Input(shape=(len(self.continuous_cols),))

        wide_linear = SparseDense(self.class_num)(input_wide)

        if (self.model_type == "wide"):
            out = Activation("softmax")(wide_linear)
            model = Model(input_wide, out)
        elif (self.model_type == "deep"):
            (input_deep, merge_list) = self._deep_merge(input_ind, input_emb, input_con)
            deep_linear = self._deep_hidden(merge_list)
            out = Activation("softmax")(deep_linear)
            model = Model(input_deep, out)
        elif (self.model_type == "wide_n_deep"):
            (input_deep, merge_list) = self._deep_merge(input_ind, input_emb, input_con)
            deep_linear = self._deep_hidden(merge_list)
            merged = merge([wide_linear, deep_linear], "sum")
            out = Activation("softmax")(merged)
            model = Model([input_wide] + input_deep, out)

        else:
            raise TypeError("Unsupported model_type: %s" % self.model_type)

        return model

    def _deep_hidden(self, merge_list):
        if (len(merge_list) == 1):
            merged = merge_list[0]
        else:
            merged = merge(merge_list, "concat")
        linear = Dense(self.hidden_layers[0], activation="relu")(merged)

        for ilayer in range(1, len(self.hidden_layers)):
            linear_mid = Dense(self.hidden_layers[ilayer], activation="relu")(linear)
            linear = linear_mid
        last = Dense(self.class_num, activation="relu")(linear)
        return last

    def _deep_merge(self, input_ind, input_emb, input_cont):
        embed_width = 0
        embed = []
        for i in range(0, len(self.embed_in_dims)):
            flat_select = Flatten()(Select(1, embed_width)(input_emb))
            iembed = Embedding(self.embed_in_dims[i] + 1, self.embed_out_dims[i],
                               init="normal")(flat_select)
            flat_embed = Flatten()(iembed)
            embed.append(flat_embed)
            embed_width = embed_width + 1

        has_ind = len(self.indicator_dims) > 0
        has_emd = len(self.embed_in_dims) > 0
        has_cont = len(self.continuous_cols) > 0
        if (has_ind and has_emd and has_cont):
            input = [input_ind, input_emb, input_cont]
            merged_list = [input_ind] + embed + [input_cont]
        elif (not has_ind and has_emd and has_cont):
            input = [input_emb, input_cont]
            merged_list = embed + [input_cont]
        elif (has_ind and (not has_emd) and has_cont):
            input = [input_ind, input_cont]
            merged_list = [input_ind, input_cont]
        elif (has_ind and has_emd and (not has_cont)):
            input = [input_ind, input_emb]
            merged_list = [input_ind] + embed
        elif ((not has_ind) and (not has_emd) and has_cont):
            input = [input_cont]
            merged_list = [input_cont]
        elif ((not has_ind) and has_emd and (not has_cont)):
            input = [input_emb]
            merged_list = embed
        elif (has_ind and (not has_emd) and not (has_cont)):
            input = [input_ind]
            merged_list = [input_ind]
        else:
            raise TypeError("Empty deep model for: %s" % self.model_type)

        return (input, merged_list)

    @staticmethod
    def load_model(path, weight_path=None, bigdl_type="float"):
        """
        Load an existing WideAndDeep model (with weights).

        # Arguments
        path: The path for the pre-defined model.
              Local file system, HDFS and Amazon S3 are supported.
              HDFS path should be like 'hdfs://[host]:[port]/xxx'.
              Amazon S3 path should be like 's3a://bucket/xxx'.
        weight_path: The path for pre-trained weights if any. Default is None.
        """
        jmodel = callZooFunc(bigdl_type, "loadWideAndDeep", path, weight_path)
        model = ZooModel._do_load(jmodel, bigdl_type)
        labor_model = KerasZooModel._do_load(jmodel, bigdl_type)
        model.model = labor_model
        model.__class__ = WideAndDeep
        return model

File: zoo/src/main/scala/com/intel/analytics/zoo/models/recommendation/WideAndDeep.scala
/*
 * Copyright 2018 Analytics Zoo Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.analytics.zoo.models.recommendation

import com.intel.analytics.bigdl.Module
import com.intel.analytics.bigdl.nn.Graph.ModuleNode
import com.intel.analytics.bigdl.nn.{CAddTable, Concat, Linear, LogSoftMax
  , LookupTable, LookupTableSparse, ParallelTable, ReLU, Sequential, Zeros}
import com.intel.analytics.bigdl.nn.abstractnn.AbstractModule
import com.intel.analytics.zoo.pipeline.api.keras.models.Model
import com.intel.analytics.bigdl.tensor.Tensor
import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
import com.intel.analytics.bigdl.utils.Shape
import com.intel.analytics.zoo.models.common.ZooModel
import com.intel.analytics.zoo.pipeline.api.keras.layers._

import scala.reflect.ClassTag

/**
 * The same data information shared by the WideAndDeep model and its feature generation part.
 *
 * @param wideBaseCols   Data of wideBaseCols together with wideCrossCols will be fed into the
 *                       wide model.
 * @param wideBaseDims   Dimensions of wideBaseCols. The dimensions of the data in wideBaseCols
 *                       should be within the range of wideBaseDims.
 * @param wideCrossCols  Data of wideCrossCols will be fed into the wide model.
 * @param wideCrossDims  Dimensions of wideCrossCols. The dimensions of the data in wideCrossCols
 *                       should be within the range of wideCrossDims.
 * @param indicatorCols  Data of indicatorCols will be fed into the deep model as multi-hot vectors.
 * @param indicatorDims  Dimensions of indicatorCols. The dimensions of the data in indicatorCols
 *                       should be within the range of indicatorDims.
 * @param embedCols      Data of embedCols will be fed into the deep model as embeddings.
 * @param embedInDims    Input dimension of the data in embedCols. The dimensions of the data
 *                       in embedCols should be within the range of embedInDims.
 * @param embedOutDims   The dimensions of embeddings.
 * @param continuousCols Data of continuousCols will be treated as continuous values
 *                       for the deep model.
 * @param label          The name of the 'label' column. Default is 'label'.
 */
case class ColumnFeatureInfo(wideBaseCols: Array[String] = Array[String](),
                             wideBaseDims: Array[Int] = Array[Int](),
                             wideCrossCols: Array[String] = Array[String](),
                             wideCrossDims: Array[Int] = Array[Int](),
                             indicatorCols: Array[String] = Array[String](),
                             indicatorDims: Array[Int] = Array[Int](),
                             embedCols: Array[String] = Array[String](),
                             embedInDims: Array[Int] = Array[Int](),
                             embedOutDims: Array[Int] = Array[Int](),
                             continuousCols: Array[String] = Array[String](),
                             label: String = "label") extends Serializable{
  override def toString: String = {
    "wideBaseCols:" + wideBaseCols.mkString(",") +"\n" +
    "wideBaseDims:" + wideBaseDims.mkString(",") + "\n" +
    "wideCrossCols:" + wideCrossCols.mkString(",") +"\n" +
    "wideCrossDims:" + wideCrossDims.mkString(",") +"\n" +
    "indicatorCols:" + indicatorCols.mkString(",") + "\n" +
    "indicatorDims:" + indicatorDims.mkString(",") +"\n" +
    "embedCols:" + embedCols.mkString(",") +"\n" +
    "embedInDims:" + embedInDims.mkString(",") +"\n" +
    "embedOutDims:" + embedOutDims.mkString(",") +"\n" +
    "continuousCols:" + continuousCols.mkString(",") +"\n" +
    "label:" + label

  }
}

/**
 * The Wide and Deep model used for recommendation.
 *
 * @param modelType      String. "wide", "deep", "wide_n_deep" are supported.
 *                       Default is "wide_n_deep".
 * @param numClasses     The number of classes. Positive integer.
 * @param wideBaseDims   Dimensions of wideBaseCols. The dimensions of the data in wideBaseCols
 *                       should be within the range of wideBaseDims.
 * @param wideCrossDims  Dimensions of crossed columns. The dimensions of the data in wideCrossCols
 *                       should be within the range of wideCrossDims.
 * @param indicatorDims  Dimensions of indicatorCols. The dimensions of the data in indicatorCols
 *                       should be within the range of indicatorDims.
 * @param embedInDims    Input dimension of the data in embedCols. The dimensions of the data
 *                       in embedCols should be within the range of embedInDims.
 * @param embedOutDims   The dimensions of embeddings.
 * @param continuousCols Data of continuousCols is treated as continuous values for the deep model.
 * @param hiddenLayers   Units of hidden layers for the deep model. Array of positive integers.
 *                       Default is Array(40, 20, 10).
 * @tparam T Numeric type of parameter(e.g. weight, bias). Only support float/double now.
 */
class WideAndDeep[T: ClassTag](
      val modelType: String = "wide_n_deep",
      val numClasses: Int,
      val wideBaseDims: Array[Int] = Array[Int](),
      val wideCrossDims: Array[Int] = Array[Int](),
      val indicatorDims: Array[Int] = Array[Int](),
      val embedInDims: Array[Int] = Array[Int](),
      val embedOutDims: Array[Int] = Array[Int](),
      val continuousCols: Array[String] = Array[String](),
      val hiddenLayers: Array[Int] = Array(40, 20, 10))(implicit ev: TensorNumeric[T])
  extends Recommender[T] {

  override def buildModel(): AbstractModule[Tensor[T], Tensor[T], T] = {

    val inputWide: ModuleNode[T] = Input(inputShape = Shape(wideBaseDims.sum + wideCrossDims.sum))
    val inputInd: ModuleNode[T] = Input(inputShape = Shape(indicatorDims.sum))
    val inputEmb = Input(inputShape = Shape(embedInDims.length))
    val inputCon = Input(inputShape = Shape(continuousCols.length))

    val wideLinear = SparseDense(numClasses).inputs(inputWide)

    modelType match {
      case "wide" =>
        val out = Activation("softmax").inputs(wideLinear)
        val model: Model[T] = Model(Array(inputWide), out)
          model.asInstanceOf[AbstractModule[Tensor[T], Tensor[T], T]]

      case "deep" =>
        val (inputDeep, mergeList) = deepMerge(inputInd, inputEmb, inputCon)
        val deepLinear = deepHidden(mergeList.toList)
        val out = Activation("softmax").inputs(deepLinear)
        Model(inputDeep, out).asInstanceOf[AbstractModule[Tensor[T], Tensor[T], T]]

      case "wide_n_deep" =>
        val (inputDeep, mergeList) = deepMerge(inputInd, inputEmb, inputCon)
        val deepLinear = deepHidden(mergeList)
        val merged = Merge.merge(List(wideLinear, deepLinear), "sum")
        val out = Activation("softmax").inputs(merged)
        Model(Array(inputWide) ++ inputDeep, out)
          .asInstanceOf[AbstractModule[Tensor[T], Tensor[T], T]]
      case _ =>
        throw new IllegalArgumentException("unknown type")
    }
  }

  private def deepHidden(mergeList: List[ModuleNode[T]]): ModuleNode[T] = {

    val merge1 = if (mergeList.size == 1) mergeList(0) else Merge.merge(mergeList, "concat")

    var linear = Dense(hiddenLayers(0), activation = "relu").inputs(merge1)

    (1 to hiddenLayers.length - 1).map(i => {
      linear = Dense(hiddenLayers(i), activation = "relu").inputs(linear)
    })

    Dense(numClasses).inputs(linear)

  }

  private def deepMerge(ind: ModuleNode[T], emb: ModuleNode[T], cont: ModuleNode[T]) :
  (Array[ModuleNode[T]], List[ModuleNode[T]]) = {

    var embedWidth = 0
    val embLookupList = (0 to embedInDims.length - 1).map(i => {
      val lookupTable = Embedding[T](embedInDims(i) + 1, embedOutDims(i), init = "normal")
      val select = Select[T](1, embedWidth).inputs(emb)
      embedWidth = embedWidth + 1
      Flatten().inputs(lookupTable.inputs(Flatten().inputs(select)))
    }).toList

    val (input, mergeList) =
      (indicatorDims.length > 0, embedInDims.length > 0, continuousCols.length > 0) match {
      case (true, true, true) =>
        (Array(ind, emb, cont), List(ind) ++ embLookupList ++ List(cont))
      case (false, true, true) =>
        (Array(emb, cont), embLookupList ++ List(cont))
      case (true, false, true) =>
        (Array(ind, cont), List(ind) ++ List(cont))
      case (true, true, false) =>
        (Array(ind, emb), List(ind) ++ embLookupList)
      case (false, true, false) =>
        (Array(emb), embLookupList)
      case (false, false, true) =>
        (Array(cont), List(cont))
      case (true, false, false) =>
        (Array(ind), List(ind))
    }
    (input, mergeList)
  }


  /**
   * Build WideAndDeep sequential model
   *
   * @return WideAndDeep sequential model
   */
  def buildSequentialModel(): AbstractModule[Tensor[T], Tensor[T], T] = {

    val fullModel = Sequential[T]()

    val wideModel = Sequential().setName("widePart")

    wideModel.add(LookupTableSparse[T](
      wideBaseDims.sum + wideCrossDims.sum, numClasses).setInitMethod(Zeros))
      .add(com.intel.analytics.bigdl.nn.CAdd(Array(numClasses)))

    // deep model
    val deepModel = Sequential[T]().setName("deepPart")
    val deepColumn = Concat[T](2)

    // add indicator encoded columns
    var indicatorWidth = 0
    if (indicatorDims.length > 0) {
      indicatorWidth = indicatorDims.sum
      deepColumn.add(Sequential[T]().add(com.intel.analytics.bigdl.nn.Narrow[T](2,
        1, indicatorWidth)))
        .setName("indicator")
    }

    // add embedding columns
    require(embedInDims.length == embedOutDims.length, s"size of embedingColumns should match")
    var embedWidth = 0
    var embedOutputDimSum = 0
    if (embedInDims.length > 0) {
      embedOutputDimSum = embedOutDims.sum
      for (i <- 0 until embedInDims.length) {
        val lookupTable = LookupTable[T](embedInDims(i), embedOutDims(i))
        lookupTable.setWeightsBias(
          Array(Tensor(embedInDims(i), embedOutDims(i)).randn(0, 0.1)))
        deepColumn.add(
          Sequential[T]().add(com.intel.analytics.bigdl.nn.Select[T](2
            , indicatorWidth + 1 + embedWidth))
            .add(lookupTable))
        embedWidth += 1
      }
    }

    // add continuous columns
    var continuousWidth = 0
    if (continuousCols.length > 0) {
      continuousWidth = continuousCols.length
      deepColumn.add(
        Sequential[T]()
          .add(com.intel.analytics.bigdl.nn.Narrow[T](2, indicatorWidth + embedWidth + 1
            , continuousWidth)))
    }

    // add hidden layers
    deepModel.add(deepColumn)
      .add(Linear[T](indicatorWidth + embedOutputDimSum + continuousWidth, hiddenLayers(0)))
      .add(ReLU())
    for (i <- 1 to hiddenLayers.length - 1) {
      deepModel.add(Linear[T](hiddenLayers(i - 1), hiddenLayers(i))).add(ReLU())
    }
    deepModel.add(Linear[T](hiddenLayers.last, numClasses))

    // create whole model
    modelType match {
      case "wide_n_deep" =>
        fullModel.add(ParallelTable().add(wideModel).add(deepModel))
          .add(CAddTable()).add(LogSoftMax())
        fullModel.asInstanceOf[AbstractModule[Tensor[T], Tensor[T], T]]
      case "wide" =>
        wideModel.add(LogSoftMax())
        wideModel.asInstanceOf[AbstractModule[Tensor[T], Tensor[T], T]]
      case "deep" =>
        deepModel.add(LogSoftMax())
        deepModel.asInstanceOf[AbstractModule[Tensor[T], Tensor[T], T]]
      case _ =>
        throw new IllegalArgumentException("unknown type")
    }
  }
}

object WideAndDeep {
  /**
   * The factory method to create a WideAndDeep instance.
   *
   * @param modelType    String. "wide", "deep", "wide_n_deep" are supported.
   *                     Default is "wide_n_deep".
   * @param numClasses   The number of classes. Positive integer.
   * @param columnInfo   An instance of ColumnFeatureInfo.
   * @param hiddenLayers Units of hidden layers for the deep model. Array of positive integers.
   *                     Default is Array(40, 20, 10).
   */
  def apply[@specialized(Float, Double) T: ClassTag](
      modelType: String = "wide_n_deep",
      numClasses: Int,
      columnInfo: ColumnFeatureInfo,
      hiddenLayers: Array[Int] = Array(40, 20, 10))
     (implicit ev: TensorNumeric[T]): WideAndDeep[T] = {
    require(columnInfo.wideBaseCols.length == columnInfo.wideBaseDims.length,
      s"size of wideBaseColumns should match")
    require(columnInfo.wideCrossCols.length == columnInfo.wideCrossDims.length,
      s"size of wideCrossColumns should match")
    require(columnInfo.indicatorCols.length == columnInfo.indicatorDims.length,
      s"size of indicatorColumns should match")
    require(columnInfo.embedCols.length == columnInfo.embedInDims.length &&
      columnInfo.embedCols.length == columnInfo.embedOutDims.length,
      s"size of embeddingColumns should match")

    new WideAndDeep[T](modelType,
      numClasses,
      columnInfo.wideBaseDims,
      columnInfo.wideCrossDims,
      columnInfo.indicatorDims,
      columnInfo.embedInDims,
      columnInfo.embedOutDims,
      columnInfo.continuousCols,
      hiddenLayers).build()
  }

  /**
   * The factory method to create a WideAndDeep Sequential Model instance.
   *
   * @param modelType String. "wide", "deep", "wide_n_deep" are supported.
   *                  Default is "wide_n_deep".
   * @param numClasses The number of classes. Positive integer.
   * @param columnInfo An instance of ColumnFeatureInfo.
   * @param hiddenLayers Units of hidden layers for the deep model. Array of positive integers.
   *                     Default is Array(40, 20, 10).
   */
  def sequential[@specialized(Float, Double) T: ClassTag](
                                                      modelType: String = "wide_n_deep",
                                                      numClasses: Int,
                                                      columnInfo: ColumnFeatureInfo,
                                                      hiddenLayers: Array[Int] = Array(40, 20, 10))
                                                    (implicit ev: TensorNumeric[T]): Module[T] = {
    require(columnInfo.wideBaseCols.length == columnInfo.wideBaseDims.length,
      s"size of wideBaseColumns should match")
    require(columnInfo.wideCrossCols.length == columnInfo.wideCrossDims.length,
      s"size of wideCrossColumns should match")
    require(columnInfo.indicatorCols.length == columnInfo.indicatorDims.length,
      s"size of indicatorColumns should match")
    require(columnInfo.embedCols.length == columnInfo.embedInDims.length &&
      columnInfo.embedCols.length == columnInfo.embedOutDims.length,
      s"size of embeddingColumns should match")

    new WideAndDeep[T](modelType,
      numClasses,
      columnInfo.wideBaseDims,
      columnInfo.wideCrossDims,
      columnInfo.indicatorDims,
      columnInfo.embedInDims,
      columnInfo.embedOutDims,
      columnInfo.continuousCols,
      hiddenLayers).buildSequentialModel()
  }

  /**
   * Load an existing WideAndDeep model (with weights).
   *
   * @param path       The path for the pre-defined model.
   *                   Local file system, HDFS and Amazon S3 are supported.
   *                   HDFS path should be like "hdfs://[host]:[port]/xxx".
   *                   Amazon S3 path should be like "s3a://bucket/xxx".
   * @param weightPath The path for pre-trained weights if any. Default is null.
   * @tparam T Numeric type of parameter(e.g. weight, bias). Only support float/double now.
   */
  def loadModel[T: ClassTag](
      path: String,
      weightPath: String = null)(implicit ev: TensorNumeric[T]): WideAndDeep[T] = {
    ZooModel.loadModel(path, weightPath).asInstanceOf[WideAndDeep[T]]
  }
}
